src/webgpu/shader/execution/robust_access.spec.ts

```diff
@@ -0,0 +1,426 @@
+    for (const matrixType of kMatrixContainerTypes) {
+      const matrixInfo = kMatrixContainerTypeInfo[matrixType];
+      yield {
+        type: `${matrixType}<${scalarType}>`,
+        _typeInfo: { elementType: `vec${matrixInfo.innerLength}<${scalarType}>`, ...matrixInfo },
+      };
+    }
+  }
+
+  const arrayTypeInfo = {
+    elementType: baseType,
+    arrayLength: kArrayLength,
+    layout: scalarInfo.layout
+      ? {
+          alignment: scalarInfo.layout.alignment,
+          size: kArrayLength * arrayStride(scalarInfo.layout),
+        }
+      : undefined,
+  };
+  // Sized array
+  yield { type: `array<${scalarType},${kArrayLength}>`, _typeInfo: arrayTypeInfo };
+  // Unsized array
+  if (storageClass === 'storage') {
+    yield { type: `array<${scalarType}>`, _typeInfo: arrayTypeInfo };
+  }
+}
+
+function arrayStride(elementLayout: { size: number; alignment: number }) {
+  return align(elementLayout.size, elementLayout.alignment);
+}
+
+g.test('linear_memory')
+  .desc(
+    `For each indexable data type (vec, mat, sized/unsized array, of various scalar types), attempts
+    to access (read, write, atomic load/store) a region of memory (buffer or internal) at various
+    (signed/unsigned) indices. Checks that the accesses conform to robust access (OOB reads only
+    return bound memory, OOB writes don't write OOB).
+
+    TODO: Test in/out storage classes.
+    TODO: Test vertex and fragment stages.
+    TODO: Test using a dynamic offset instead of a static offset into uniform/storage bindings.
+  `
+  )
+  .params(u =>
+    u
+      .combineWithParams([
+        { storageClass: 'storage', storageMode: 'read', access: 'read' },
+        { storageClass: 'storage', storageMode: 'write', access: 'write' },
+        { storageClass: 'storage', storageMode: 'read_write', access: 'read' },
+        { storageClass: 'storage', storageMode: 'read_write', access: 'write' },
+        { storageClass: 'uniform', access: 'read' },
+        { storageClass: 'private', access: 'read' },
+        { storageClass: 'private', access: 'write' },
+        { storageClass: 'function', access: 'read' },
+        { storageClass: 'function', access: 'write' },
+        { storageClass: 'workgroup', access: 'read' },
+        { storageClass: 'workgroup', access: 'write' },
+      ] as const)
+      .combine('atomic', [false, true])
+      .expand('baseType', function* (p) {
+        yield* ['i32', 'u32'] as const;
+        if (!p.atomic) yield* ['f32', 'bool'] as const;
+      })
+      .filter(p => {
+        const info = kScalarTypeInfo[p.baseType];
+
+        // Storage and uniform require host-sharable types.
+        const isHostShared = p.storageClass === 'storage' || p.storageClass === 'uniform';
+        if (isHostShared && info.layout === undefined) return false;
+
+        // Atomic access requires atomic type and storage/workgroup memory.
+        const classSupportsAtomics = p.storageClass === 'storage' || p.storageClass === 'workgroup';
```

Comment:
Maybe this check could be in the `.combine('atomic')` that would instead become an expand.